{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import *\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_codon_table = [\n",
    "        (\"ATG\",\"M/Start\"), (\"ATA\",\"I\"), (\"ATC\",\"I\"), (\"ATT\",\"I\"), (\"ACG\",\"T\"), (\"ACA\",\"T\"),\n",
    "        (\"ACC\",\"T\"), (\"ACT\", \"T\"), (\"AAG\",\"K\"), (\"AAA\",\"K\"), (\"AAC\",\"N\"), (\"AAT\",\"N\"),\n",
    "        (\"AGG\",\"R\"), (\"AGA\",\"R\"), (\"AGC\",\"S\"), (\"AGT\",\"S\"), \n",
    "        (\"GTG\",\"V\"), (\"GTA\",\"V\"), (\"GTC\",\"V\"), (\"GTT\",\"V\"), (\"GCG\",\"A\"), (\"GCA\",\"A\"), (\"GCC\",\"A\"), (\"GCT\",\"A\"), \n",
    "        (\"GAG\",\"E\"), (\"GAA\",\"E\"), (\"GAC\",\"D\"), (\"GAT\",\"D\"), (\"GGG\",\"G\"), (\"GGA\",\"G\"), (\"GGC\",\"G\"), (\"GGT\",\"G\"), \n",
    "        (\"CTG\",\"L\"), (\"CTA\",\"L\"), (\"CTC\",\"L\"), (\"CTT\",\"L\"), (\"CCG\",\"P\"), (\"CCA\",\"P\"), (\"CCC\",\"P\"), (\"CCT\",\"P\"),\n",
    "        (\"CAG\",\"Q\"), (\"CAA\",\"Q\"), (\"CAC\",\"H\"), (\"CAT\",\"H\"), (\"CGG\",\"R\"), (\"CGA\",\"R\"), (\"CGC\",\"R\"), (\"CGT\",\"R\"),\n",
    "        (\"TTG\",\"L\"), (\"TTA\",\"L\"), (\"TTC\",\"F\"), (\"TTT\",\"F\"), (\"TCG\",\"S\"), (\"TCA\",\"S\"), (\"TCC\",\"S\"), (\"TCT\",\"S\"),\n",
    "        (\"TAC\",\"Y\"), (\"TAT\",\"Y\"), (\"TGG\",\"W\"), (\"TGC\",\"C\"), (\"TGT\",\"C\"), (\"TAG\",\"Amber/Stop\"),\n",
    "        (\"TAA\",\"Ochre/Stop\"),(\"TGA\",\"Opal/Stop\")\n",
    "]\n",
    "\n",
    "Nucl = [\"A\", \"T\", \"G\", \"C\"]\n",
    "\n",
    "stop = [\"TAG\", \"TAA\", \"TGA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rules:\n",
    "# -1 = default\n",
    "# 0 - 4 = exclude Nucleotide\n",
    "def create_protein(seq_length, codon_excl, rule=-1, rand_pos_pick=-1, AminExcl=0): #AminExcl -# amino acids to exclude\n",
    "    protein = []\n",
    "    tmp = []\n",
    "    Excl = [] # \" \"\n",
    "    #if AminExcl == 1:\n",
    "    #    Excl = DNA_codon_table[randint(1, 61)][1]\n",
    "    tmp_table = DNA_codon_table[1:61]\n",
    "    limit_table = []\n",
    "    #print(tmp_table)\n",
    "\n",
    "    if AminExcl > 0: # > 1\n",
    "        #Excl = []\n",
    "        for i in range(AminExcl):\n",
    "            #print(len(tmp_table))\n",
    "            tmp = tmp_table[randint(0, len(tmp_table)-1)]\n",
    "            #while(tmp[1] in Excl):\n",
    "                #print(tmp)\n",
    "                #tmp_table.remove(tmp)\n",
    "                #tmp = tmp_table[randint(0, len(tmp_table)-1)]\n",
    "            Excl.append(tmp[1])\n",
    "            tmp_table.remove(tmp)\n",
    "            \n",
    "    for i in range(1, 60):\n",
    "        if(DNA_codon_table[i][1] not in codon_excl):\n",
    "            limit_table.append(DNA_codon_table[i])\n",
    "            \n",
    "    #print(\"Amino Acid exclusion: \", Excl)\n",
    "    triplet = limit_table[randint(0, len(limit_table)-1)]\n",
    "    for j in range(0, seq_length):  #generating one specific protein sequence\n",
    "        if(rule == -1):\n",
    "            #protein.append(aminos[randint(0, len(aminos)-1)])  #DNA sequences\n",
    "            if(rand_pos_pick > -1):\n",
    "                if(rand_pos_pick == j):\n",
    "                    tmp = DNA_codon_table[randint(1, 60)]\n",
    "                    while(tmp[1] == triplet[1]):\n",
    "                        tmp = DNA_codon_table[randint(1, 60)]\n",
    "                    protein.append(tmp[0])\n",
    "                else:\n",
    "                    tmp = limit_table[randint(0, len(limit_table)-1)]\n",
    "                    while (tmp[1] != triplet[1]):\n",
    "                        tmp = limit_table[randint(0, len(limit_table)-1)]\n",
    "                    protein.append(tmp[0])\n",
    "            else:\n",
    "                protein.append(triplet[0])\n",
    "            \n",
    "        else:\n",
    "            tmp = DNA_codon_table[randint(1, 60)][0]\n",
    "            #aminos[randint(0, len(aminos)-1)]\n",
    "            while(Nucl[rule % 4] in tmp):\n",
    "                tmp = DNA_codon_table[randint(1, 60)][0]\n",
    "                #aminos[randint(0, len(aminos)-1)]\n",
    "            protein.append(tmp)\n",
    "    return protein, Excl, triplet[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_comb(val, codon_dict):\n",
    "    pos_comb = 0\n",
    "    for key, value in codon_dict.items():\n",
    "        if value == val:\n",
    "            pos_comb += 1\n",
    "    return pos_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_comb(val, codon_dict, rule=-1):\n",
    "    combs = []\n",
    "    for key, value in codon_dict.items():\n",
    "        if value == val:\n",
    "            if(rule != -1):\n",
    "                if(not Nucl[rule % 4] in key):\n",
    "                    combs.append(key)\n",
    "            else:\n",
    "                combs.append(key)\n",
    "    return combs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sequences(protein, num_seq, rule=-1, rand_change=0):\n",
    "    '''\n",
    "        1. Protein sequence\n",
    "        2. used_key lists for each amino acid\n",
    "                           -OR-\n",
    "           lists of possible key values for each key\n",
    "    '''\n",
    "    sequence = []\n",
    "    tmp = []\n",
    "    amino_list = []\n",
    "    protein_len = len(protein)\n",
    "    for i in range(0, protein_len):\n",
    "        amino_list.append(False)\n",
    "    amino_list[0] = True\n",
    "    codon_dict = dict(DNA_codon_table)\n",
    "    #print(codon_dict)\n",
    "    seq_count = 0\n",
    "    comb_dict = {}\n",
    "    #comb_list = []\n",
    "    for i in range(0, len(protein)):\n",
    "        #print(codon_dict[protein[i]])\n",
    "        combs = find_comb(codon_dict[protein[i]], codon_dict, rule)\n",
    "        #print(type(comb_dict))\n",
    "        comb_dict[i] = combs\n",
    "        #comb_list.append(combs)\n",
    "    #print(comb_dict)\n",
    "    index = 0\n",
    "    comb_tracker = [0 for i in range(0,len(protein))]\n",
    "    #print(comb_tracker)\n",
    "\n",
    "    while seq_count < num_seq:\n",
    "        for i in range(0, protein_len):\n",
    "            tmp.append(choice(comb_dict[i]))\n",
    "        sequence.append(list(tmp))\n",
    "        tmp = []\n",
    "        seq_count += 1\n",
    "        \n",
    "    #print(sequence)\n",
    "    #print(\"amount of sequences: \", len(sequence))\n",
    "        \n",
    "    if rand_change > 0:\n",
    "        rand_samples = randint(1, num_seq-1)  #alter rand_samples amount of sequences in protein class\n",
    "        #print(\"rand_samples: \", rand_samples)\n",
    "        for i in range(rand_samples):\n",
    "            which_samp = randint(0, num_seq-1) #which seq to modify\n",
    "            #print(\"Which samp: \", which_samp)\n",
    "            for j in range(rand_change):   #which amino acids to modify\n",
    "                loc = randint(0, protein_len-1)\n",
    "                #print(\"location: \", loc)\n",
    "                sequence[which_samp][loc] = DNA_codon_table[randint(1, 60)][0]\n",
    "            \n",
    "            \n",
    "    return sequence\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_set(num_aminos, seq_length, num_classes, seq_per_class):\n",
    "    s = \"\"\n",
    "    if num_aminos < 3:\n",
    "        print(\"Need more aminos\")\n",
    "        return\n",
    "    rand_aminos = []\n",
    "    classes = []\n",
    "    seq = []\n",
    "    Excl = []\n",
    "    codonExcl = []\n",
    "    #for i in range(1, num_aminos-3): # -2 because of start and stop codons\n",
    "    #    rand_aminos.append(DNA_codon_table[randint(1, 62)][0])  #looking at DNA\n",
    "\n",
    "    for i in range(0, num_classes):  #generating diffferent proteins\n",
    "        protein, tmp, CExcl = create_protein(seq_length, codonExcl, -1, rand_pos_pick=(seq_length-1))\n",
    "        Excl.append(tmp)\n",
    "        codonExcl.append(CExcl)\n",
    "        print(\"protein \", protein)\n",
    "        seq_list = create_sequences(protein, seq_per_class)\n",
    "        for j in seq_list:\n",
    "        #    j.insert(0, \"ATG\")\n",
    "        #    j.append(choice(stop))\n",
    "            seq.append(''.join(j))\n",
    "            classes.append(i)\n",
    "            #print(j)\n",
    "        #print(type(seq_new))\n",
    "        print(\"Finished class: \", i+1)\n",
    "    #print(seq)\n",
    "    print(codonExcl)\n",
    "    return classes, seq, Excl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "protein  ['GGT', 'GGC', 'GGT', 'GGA', 'GGG', 'GGT', 'GGG', 'GGG', 'GGC', 'GGA', 'GGG', 'GGC', 'GGG', 'GGG', 'GGT', 'GGT', 'GGC', 'GGG', 'GGA', 'GGC', 'GGA', 'GGG', 'GGT', 'GGG', 'GGA', 'GGC', 'GGG', 'GGA', 'GGT', 'GGT', 'GGT', 'GGA', 'GGC', 'GGC', 'GGG', 'GGG', 'GGG', 'GGT', 'GGT', 'GGT', 'GGA', 'GGC', 'GGC', 'GGG', 'GGC', 'GGA', 'GGC', 'GGA', 'GGA', 'GAC']\n",
      "Finished class:  1\n",
      "protein  ['CTA', 'CTC', 'CTC', 'CTA', 'TTG', 'CTC', 'CTT', 'CTT', 'CTA', 'CTC', 'CTC', 'CTA', 'CTG', 'CTC', 'CTC', 'TTA', 'TTA', 'TTA', 'TTG', 'CTA', 'CTT', 'CTG', 'TTA', 'CTG', 'TTA', 'TTG', 'TTG', 'CTT', 'CTC', 'CTT', 'CTA', 'TTA', 'CTT', 'TTA', 'TTG', 'CTT', 'TTG', 'CTG', 'CTT', 'CTA', 'TTA', 'CTG', 'TTA', 'CTA', 'TTA', 'CTA', 'CTC', 'TTG', 'CTG', 'CGC']\n",
      "Finished class:  2\n",
      "protein  ['GTG', 'GTA', 'GTG', 'GTG', 'GTG', 'GTA', 'GTG', 'GTT', 'GTT', 'GTT', 'GTC', 'GTT', 'GTC', 'GTA', 'GTC', 'GTG', 'GTA', 'GTT', 'GTT', 'GTT', 'GTT', 'GTC', 'GTC', 'GTT', 'GTA', 'GTG', 'GTG', 'GTC', 'GTA', 'GTC', 'GTC', 'GTC', 'GTG', 'GTA', 'GTA', 'GTC', 'GTC', 'GTT', 'GTA', 'GTG', 'GTT', 'GTT', 'GTA', 'GTG', 'GTG', 'GTC', 'GTC', 'GTT', 'GTC', 'ATC']\n",
      "Finished class:  3\n",
      "protein  ['GCG', 'GCA', 'GCA', 'GCC', 'GCT', 'GCT', 'GCT', 'GCA', 'GCA', 'GCT', 'GCC', 'GCA', 'GCC', 'GCA', 'GCC', 'GCG', 'GCT', 'GCT', 'GCT', 'GCA', 'GCC', 'GCC', 'GCA', 'GCA', 'GCA', 'GCG', 'GCA', 'GCT', 'GCA', 'GCC', 'GCC', 'GCG', 'GCA', 'GCT', 'GCA', 'GCA', 'GCC', 'GCT', 'GCT', 'GCA', 'GCT', 'GCG', 'GCG', 'GCG', 'GCC', 'GCT', 'GCC', 'GCC', 'GCT', 'AAA']\n",
      "Finished class:  4\n",
      "protein  ['ACA', 'ACT', 'ACG', 'ACC', 'ACA', 'ACC', 'ACA', 'ACG', 'ACA', 'ACC', 'ACT', 'ACA', 'ACC', 'ACT', 'ACG', 'ACC', 'ACA', 'ACC', 'ACT', 'ACC', 'ACC', 'ACA', 'ACC', 'ACA', 'ACA', 'ACT', 'ACA', 'ACT', 'ACC', 'ACA', 'ACG', 'ACT', 'ACG', 'ACG', 'ACA', 'ACG', 'ACG', 'ACT', 'ACA', 'ACC', 'ACT', 'ACT', 'ACC', 'ACC', 'ACG', 'ACA', 'ACT', 'ACC', 'ACG', 'GCC']\n",
      "Finished class:  5\n",
      "protein  ['ATA', 'ATA', 'ATT', 'ATA', 'ATC', 'ATT', 'ATC', 'ATT', 'ATT', 'ATT', 'ATT', 'ATT', 'ATA', 'ATC', 'ATC', 'ATT', 'ATA', 'ATC', 'ATC', 'ATT', 'ATA', 'ATC', 'ATA', 'ATC', 'ATT', 'ATA', 'ATC', 'ATA', 'ATT', 'ATA', 'ATT', 'ATA', 'ATA', 'ATA', 'ATT', 'ATT', 'ATC', 'ATT', 'ATT', 'ATA', 'ATT', 'ATT', 'ATT', 'ATC', 'ATA', 'ATA', 'ATA', 'ATC', 'ATA', 'CAG']\n",
      "Finished class:  6\n",
      "protein  ['TCT', 'TCG', 'TCC', 'TCG', 'AGT', 'TCT', 'TCT', 'AGC', 'TCG', 'AGC', 'AGT', 'AGT', 'AGT', 'AGC', 'TCC', 'AGT', 'TCA', 'AGT', 'AGC', 'AGC', 'TCA', 'TCG', 'TCA', 'TCG', 'TCA', 'TCC', 'AGT', 'TCA', 'AGC', 'TCG', 'AGT', 'TCT', 'AGT', 'AGT', 'TCG', 'AGC', 'TCC', 'TCG', 'AGT', 'TCA', 'TCT', 'TCA', 'AGT', 'TCA', 'TCC', 'AGC', 'TCG', 'AGT', 'AGT', 'ATC']\n",
      "Finished class:  7\n",
      "protein  ['TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'TGC', 'GAT']\n",
      "Finished class:  8\n",
      "protein  ['AAT', 'AAT', 'AAT', 'AAC', 'AAT', 'AAT', 'AAC', 'AAC', 'AAT', 'AAC', 'AAC', 'AAT', 'AAT', 'AAC', 'AAT', 'AAC', 'AAC', 'AAC', 'AAT', 'AAC', 'AAC', 'AAT', 'AAC', 'AAT', 'AAT', 'AAC', 'AAT', 'AAC', 'AAT', 'AAC', 'AAC', 'AAC', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAT', 'AAC', 'AAT', 'AAC', 'AAT', 'AAT', 'AAT', 'GGG']\n",
      "Finished class:  9\n",
      "protein  ['CCC', 'CCA', 'CCA', 'CCG', 'CCA', 'CCA', 'CCC', 'CCA', 'CCG', 'CCG', 'CCA', 'CCA', 'CCG', 'CCA', 'CCC', 'CCA', 'CCT', 'CCT', 'CCA', 'CCC', 'CCC', 'CCT', 'CCG', 'CCA', 'CCG', 'CCG', 'CCC', 'CCA', 'CCA', 'CCT', 'CCT', 'CCT', 'CCG', 'CCT', 'CCG', 'CCC', 'CCG', 'CCC', 'CCC', 'CCT', 'CCC', 'CCC', 'CCG', 'CCC', 'CCT', 'CCC', 'CCT', 'CCG', 'CCT', 'TTA']\n",
      "Finished class:  10\n",
      "protein  ['AAG', 'AAG', 'AAG', 'AAA', 'AAA', 'AAG', 'AAG', 'AAA', 'AAA', 'AAG', 'AAA', 'AAA', 'AAG', 'AAG', 'AAG', 'AAG', 'AAA', 'AAA', 'AAG', 'AAG', 'AAG', 'AAG', 'AAA', 'AAG', 'AAA', 'AAG', 'AAA', 'AAG', 'AAG', 'AAA', 'AAG', 'AAG', 'AAG', 'AAG', 'AAG', 'AAA', 'AAG', 'AAG', 'AAA', 'AAA', 'AAA', 'AAG', 'AAA', 'AAG', 'AAA', 'AAG', 'AAG', 'AAA', 'AAG', 'CCT']\n",
      "Finished class:  11\n",
      "protein  ['AGA', 'CGG', 'CGG', 'AGG', 'AGA', 'CGG', 'CGG', 'CGG', 'CGA', 'CGT', 'CGC', 'AGG', 'CGC', 'AGA', 'AGG', 'CGC', 'CGC', 'CGT', 'CGA', 'CGG', 'CGG', 'CGG', 'AGG', 'AGA', 'CGG', 'CGC', 'CGA', 'CGG', 'CGA', 'AGA', 'CGT', 'CGG', 'AGG', 'CGA', 'AGG', 'AGG', 'CGG', 'CGC', 'CGT', 'CGG', 'CGT', 'CGT', 'CGC', 'CGT', 'AGG', 'AGA', 'CGA', 'CGT', 'AGA', 'CCA']\n",
      "Finished class:  12\n",
      "protein  ['GAG', 'GAG', 'GAG', 'GAA', 'GAG', 'GAA', 'GAA', 'GAG', 'GAA', 'GAA', 'GAA', 'GAA', 'GAA', 'GAG', 'GAG', 'GAG', 'GAA', 'GAG', 'GAG', 'GAA', 'GAG', 'GAA', 'GAG', 'GAG', 'GAG', 'GAG', 'GAG', 'GAG', 'GAG', 'GAG', 'GAG', 'GAA', 'GAA', 'GAG', 'GAG', 'GAG', 'GAA', 'GAA', 'GAG', 'GAA', 'GAG', 'GAG', 'GAG', 'GAA', 'GAG', 'GAA', 'GAA', 'GAG', 'GAA', 'CGA']\n",
      "Finished class:  13\n",
      "protein  ['TTC', 'TTC', 'TTC', 'TTC', 'TTC', 'TTC', 'TTT', 'TTC', 'TTT', 'TTT', 'TTC', 'TTC', 'TTT', 'TTT', 'TTC', 'TTC', 'TTT', 'TTT', 'TTT', 'TTT', 'TTT', 'TTC', 'TTT', 'TTC', 'TTT', 'TTC', 'TTT', 'TTT', 'TTT', 'TTC', 'TTT', 'TTC', 'TTC', 'TTC', 'TTC', 'TTT', 'TTT', 'TTT', 'TTT', 'TTC', 'TTT', 'TTC', 'TTC', 'TTT', 'TTT', 'TTT', 'TTC', 'TTC', 'TTC', 'GCA']\n",
      "Finished class:  14\n",
      "protein  ['GAT', 'GAC', 'GAC', 'GAT', 'GAC', 'GAC', 'GAC', 'GAT', 'GAC', 'GAC', 'GAT', 'GAC', 'GAT', 'GAT', 'GAT', 'GAT', 'GAT', 'GAT', 'GAT', 'GAC', 'GAT', 'GAT', 'GAC', 'GAC', 'GAC', 'GAT', 'GAC', 'GAC', 'GAC', 'GAT', 'GAT', 'GAC', 'GAT', 'GAT', 'GAC', 'GAT', 'GAT', 'GAT', 'GAC', 'GAC', 'GAC', 'GAC', 'GAT', 'GAT', 'GAT', 'GAT', 'GAC', 'GAT', 'GAC', 'GGC']\n",
      "Finished class:  15\n",
      "protein  ['TAT', 'TAC', 'TAT', 'TAT', 'TAC', 'TAT', 'TAT', 'TAT', 'TAC', 'TAT', 'TAC', 'TAC', 'TAC', 'TAT', 'TAC', 'TAC', 'TAC', 'TAT', 'TAC', 'TAC', 'TAC', 'TAT', 'TAT', 'TAC', 'TAT', 'TAC', 'TAT', 'TAC', 'TAT', 'TAT', 'TAT', 'TAC', 'TAC', 'TAT', 'TAT', 'TAT', 'TAC', 'TAT', 'TAC', 'TAC', 'TAT', 'TAC', 'TAT', 'TAT', 'TAT', 'TAT', 'TAC', 'TAT', 'TAC', 'GGT']\n",
      "Finished class:  16\n",
      "protein  ['TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'TGG', 'AGT']\n",
      "Finished class:  17\n",
      "protein  ['CAA', 'CAA', 'CAG', 'CAG', 'CAA', 'CAG', 'CAA', 'CAA', 'CAG', 'CAG', 'CAA', 'CAG', 'CAG', 'CAG', 'CAA', 'CAA', 'CAA', 'CAG', 'CAA', 'CAG', 'CAA', 'CAA', 'CAG', 'CAA', 'CAA', 'CAA', 'CAA', 'CAG', 'CAA', 'CAG', 'CAG', 'CAG', 'CAG', 'CAG', 'CAG', 'CAG', 'CAA', 'CAG', 'CAA', 'CAA', 'CAG', 'CAA', 'CAG', 'CAA', 'CAA', 'CAA', 'CAA', 'CAA', 'CAG', 'CCT']\n",
      "Finished class:  18\n",
      "protein  ['CAC', 'CAT', 'CAC', 'CAC', 'CAT', 'CAT', 'CAC', 'CAC', 'CAT', 'CAC', 'CAT', 'CAT', 'CAC', 'CAT', 'CAC', 'CAC', 'CAT', 'CAT', 'CAC', 'CAT', 'CAC', 'CAC', 'CAC', 'CAT', 'CAC', 'CAT', 'CAT', 'CAT', 'CAC', 'CAC', 'CAT', 'CAC', 'CAT', 'CAC', 'CAT', 'CAT', 'CAC', 'CAT', 'CAC', 'CAT', 'CAC', 'CAC', 'CAT', 'CAC', 'CAC', 'CAT', 'CAC', 'CAT', 'CAC', 'CCG']\n",
      "Finished class:  19\n",
      "['G', 'L', 'V', 'A', 'T', 'I', 'S', 'C', 'N', 'P', 'K', 'R', 'E', 'F', 'D', 'Y', 'W', 'Q', 'H']\n",
      "9500\n",
      "data64_50_19_500_same_last\n"
     ]
    }
   ],
   "source": [
    "num_aminos = 64\n",
    "seq_length = 50\n",
    "num_classes = 19\n",
    "seq_per_class = 500\n",
    "keys, val, Excl = create_set(num_aminos, seq_length, num_classes, seq_per_class)\n",
    "print(len(val))\n",
    "#print(keys)\n",
    "#print(len(val))\n",
    "#list of tuples (x,y) where x is a string and y is index\n",
    "fout = str(num_aminos) + '_' + str(seq_length) + '_' + str(num_classes) + '_' + str(seq_per_class) + '_same_last'\n",
    "print(\"data\"+fout)\n",
    "#for i in range(len(Excl)):\n",
    "    #print(\"Amino Acid exclusion \", i+1, \": \", Excl[i])\n",
    "with open('../data/data%s.csv' % fout, 'w') as f:\n",
    "#with open('data/64_30_30_10.csv', 'w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['class', 'sequence'])\n",
    "    for i in range(0, len(keys)):\n",
    "        #print(\"i: \", i)\n",
    "        writer.writerow([keys[i], val[i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
