{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import *\n",
    "import csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "DNA_codon_table = [\n",
    "        (\"ATG\",\"M/Start\"), (\"ATA\",\"I\"), (\"ATC\",\"I\"), (\"ATT\",\"I\"), (\"ACG\",\"T\"), (\"ACA\",\"T\"),\n",
    "        (\"ACC\",\"T\"), (\"ACT\", \"T\"), (\"AAG\",\"K\"), (\"AAA\",\"K\"), (\"AAC\",\"N\"), (\"AAT\",\"N\"),\n",
    "        (\"AGG\",\"R\"), (\"AGA\",\"R\"), (\"AGC\",\"S\"), (\"AGT\",\"S\"), \n",
    "        (\"GTG\",\"V\"), (\"GTA\",\"V\"), (\"GTC\",\"V\"), (\"GTT\",\"V\"), (\"GCG\",\"A\"), (\"GCA\",\"A\"), (\"GCC\",\"A\"), (\"GCT\",\"A\"), \n",
    "        (\"GAG\",\"E\"), (\"GAA\",\"E\"), (\"GAC\",\"D\"), (\"GAT\",\"D\"), (\"GGG\",\"G\"), (\"GGA\",\"G\"), (\"GGC\",\"G\"), (\"GGT\",\"G\"), \n",
    "        (\"CTG\",\"L\"), (\"CTA\",\"L\"), (\"CTC\",\"L\"), (\"CTT\",\"L\"), (\"CCG\",\"P\"), (\"CCA\",\"P\"), (\"CCC\",\"P\"), (\"CCT\",\"P\"),\n",
    "        (\"CAG\",\"Q\"), (\"CAA\",\"Q\"), (\"CAC\",\"H\"), (\"CAT\",\"H\"), (\"CGG\",\"R\"), (\"CGA\",\"R\"), (\"CGC\",\"R\"), (\"CGT\",\"R\"),\n",
    "        (\"TTG\",\"L\"), (\"TTA\",\"L\"), (\"TTC\",\"F\"), (\"TTT\",\"F\"), (\"TCG\",\"S\"), (\"TCA\",\"S\"), (\"TCC\",\"S\"), (\"TCT\",\"S\"),\n",
    "        (\"TAC\",\"Y\"), (\"TAT\",\"Y\"), (\"TGG\",\"W\"), (\"TGC\",\"C\"), (\"TGT\",\"C\"), (\"TAG\",\"Amber/Stop\"),\n",
    "        (\"TAA\",\"Ochre/Stop\"),(\"TGA\",\"Opal/Stop\")\n",
    "]\n",
    "\n",
    "Nucl = [\"A\", \"T\", \"G\", \"C\"]\n",
    "\n",
    "stop = [\"TAG\", \"TAA\", \"TGA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(DNA_codon_table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rules:\n",
    "# -1 = default\n",
    "# 0 - 4 = exclude Nucleotide\n",
    "def create_protein(seq_length, rule=-1, AminExcl=0): #AminExcl -# amino acids to exclude\n",
    "    protein = []\n",
    "    tmp = []\n",
    "    Excl = [] # \" \"\n",
    "    #if AminExcl == 1:\n",
    "    #    Excl = DNA_codon_table[randint(1, 61)][1]\n",
    "    \n",
    "    if AminExcl > 0: # > 1\n",
    "        #Excl = []\n",
    "        for i in range(AminExcl):\n",
    "            tmp = DNA_codon_table[randint(1, 61)][1]\n",
    "            while(tmp in Excl):\n",
    "                tmp = DNA_codon_table[randint(1, 61)][1]\n",
    "            Excl.append(tmp)\n",
    "            \n",
    "    #print(\"Amino Acid exclusion: \", Excl)\n",
    "    for j in range(0, seq_length):  #generating one specific protein sequence\n",
    "        if(rule == -1):\n",
    "            #protein.append(aminos[randint(0, len(aminos)-1)])  #DNA sequences\n",
    "            triplet = DNA_codon_table[randint(1, 61)]\n",
    "            if AminExcl == 1:\n",
    "                while triplet[1] == Excl:\n",
    "                    triplet = DNA_codon_table[randint(1, 61)]\n",
    "            \n",
    "            if AminExcl > 1:\n",
    "                while triplet[1] in Excl:\n",
    "                    triplet = DNA_codon_table[randint(1, 61)]\n",
    "                    \n",
    "            protein.append(triplet[0])\n",
    "        else:\n",
    "            tmp = DNA_codon_table[randint(1, 61)][0]\n",
    "            #aminos[randint(0, len(aminos)-1)]\n",
    "            while(Nucl[rule % 4] in tmp):\n",
    "                tmp = DNA_codon_table[randint(1, 61)][0]\n",
    "                #aminos[randint(0, len(aminos)-1)]\n",
    "            protein.append(tmp)\n",
    "    return protein, Excl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def num_comb(val, codon_dict):\n",
    "    pos_comb = 0\n",
    "    for key, value in codon_dict.items():\n",
    "        if value == val:\n",
    "            pos_comb += 1\n",
    "    return pos_comb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_comb(val, codon_dict, rule=-1):\n",
    "    combs = []\n",
    "    for key, value in codon_dict.items():\n",
    "        if value == val:\n",
    "            if(rule != -1):\n",
    "                if(not Nucl[rule % 4] in key):\n",
    "                    combs.append(key)\n",
    "            else:\n",
    "                combs.append(key)\n",
    "    return combs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_sequences(protein, num_seq, rule=-1, rand_change=0):\n",
    "    '''\n",
    "        1. Protein sequence\n",
    "        2. used_key lists for each amino acid\n",
    "                           -OR-\n",
    "           lists of possible key values for each key\n",
    "    '''\n",
    "    sequence = []\n",
    "    tmp = []\n",
    "    amino_list = []\n",
    "    protein_len = len(protein)\n",
    "    for i in range(0, protein_len):\n",
    "        amino_list.append(False)\n",
    "    amino_list[0] = True\n",
    "    codon_dict = dict(DNA_codon_table)\n",
    "    #print(codon_dict)\n",
    "    seq_count = 0\n",
    "    comb_dict = {}\n",
    "    #comb_list = []\n",
    "    for i in range(0, len(protein)):\n",
    "        #print(codon_dict[protein[i]])\n",
    "        combs = find_comb(codon_dict[protein[i]], codon_dict, rule)\n",
    "        #print(type(comb_dict))\n",
    "        comb_dict[i] = combs\n",
    "        #comb_list.append(combs)\n",
    "    #print(comb_dict)\n",
    "    index = 0\n",
    "    comb_tracker = [0 for i in range(0,len(protein))]\n",
    "    #print(comb_tracker)\n",
    "\n",
    "    while seq_count < num_seq:\n",
    "        for i in range(0, protein_len):\n",
    "            tmp.append(choice(comb_dict[i]))\n",
    "        sequence.append(list(tmp))\n",
    "        tmp = []\n",
    "        seq_count += 1\n",
    "        \n",
    "    #print(sequence)\n",
    "    #print(\"amount of sequences: \", len(sequence))\n",
    "        \n",
    "    if rand_change > 0:\n",
    "        rand_samples = randint(1, num_seq-1)  #alter rand_samples amount of sequences in protein class\n",
    "        #print(\"rand_samples: \", rand_samples)\n",
    "        for i in range(rand_samples):\n",
    "            which_samp = randint(0, num_seq-1) #which seq to modify\n",
    "            #print(\"Which samp: \", which_samp)\n",
    "            for j in range(rand_change):   #which amino acids to modify\n",
    "                loc = randint(0, protein_len-1)\n",
    "                #print(\"location: \", loc)\n",
    "                sequence[which_samp][loc] = DNA_codon_table[randint(1, 61)][0]\n",
    "            \n",
    "            \n",
    "    return sequence\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_set(num_aminos, seq_length, num_classes, seq_per_class):\n",
    "    s = \"\"\n",
    "    if num_aminos < 3:\n",
    "        print(\"Need more aminos\")\n",
    "        return\n",
    "    rand_aminos = []\n",
    "    classes = []\n",
    "    seq = []\n",
    "    Excl = []\n",
    "    #for i in range(1, num_aminos-3): # -2 because of start and stop codons\n",
    "    #    rand_aminos.append(DNA_codon_table[randint(1, 62)][0])  #looking at DNA\n",
    "\n",
    "    for i in range(0, num_classes):  #generating diffferent proteins\n",
    "        protein, tmp = create_protein(seq_length, -1, 15)\n",
    "        Excl.append(tmp)\n",
    "        #print(\"protein \", protein)\n",
    "        seq_list = create_sequences(protein, seq_per_class, -1, 1)\n",
    "        for j in seq_list:\n",
    "            j.insert(0, \"ATG\")\n",
    "            j.append(choice(stop))\n",
    "            seq.append(''.join(j))\n",
    "            classes.append(i)\n",
    "            #print(j)\n",
    "        #print(type(seq_new))\n",
    "        print(\"Finished class: \", i+1)\n",
    "    #print(seq)\n",
    "    return classes, seq, Excl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Finished class:  1\n",
      "Finished class:  2\n",
      "Finished class:  3\n",
      "Finished class:  4\n",
      "Finished class:  5\n",
      "Finished class:  6\n",
      "Finished class:  7\n",
      "Finished class:  8\n",
      "Finished class:  9\n",
      "Finished class:  10\n",
      "Finished class:  11\n",
      "Finished class:  12\n",
      "Finished class:  13\n",
      "Finished class:  14\n",
      "Finished class:  15\n",
      "Finished class:  16\n",
      "Finished class:  17\n",
      "Finished class:  18\n",
      "Finished class:  19\n",
      "Finished class:  20\n",
      "Finished class:  21\n",
      "Finished class:  22\n",
      "Finished class:  23\n",
      "Finished class:  24\n",
      "Finished class:  25\n",
      "Finished class:  26\n",
      "Finished class:  27\n",
      "Finished class:  28\n",
      "Finished class:  29\n",
      "Finished class:  30\n",
      "Finished class:  31\n",
      "Finished class:  32\n",
      "Finished class:  33\n",
      "Finished class:  34\n",
      "Finished class:  35\n",
      "Finished class:  36\n",
      "Finished class:  37\n",
      "Finished class:  38\n",
      "Finished class:  39\n",
      "Finished class:  40\n",
      "Finished class:  41\n",
      "Finished class:  42\n",
      "Finished class:  43\n",
      "Finished class:  44\n",
      "Finished class:  45\n",
      "Finished class:  46\n",
      "Finished class:  47\n",
      "Finished class:  48\n",
      "Finished class:  49\n",
      "Finished class:  50\n",
      "Finished class:  51\n",
      "Finished class:  52\n",
      "Finished class:  53\n",
      "Finished class:  54\n",
      "Finished class:  55\n",
      "Finished class:  56\n",
      "Finished class:  57\n",
      "Finished class:  58\n",
      "Finished class:  59\n",
      "Finished class:  60\n",
      "Finished class:  61\n",
      "Finished class:  62\n",
      "Finished class:  63\n",
      "Finished class:  64\n",
      "Finished class:  65\n",
      "Finished class:  66\n",
      "Finished class:  67\n",
      "Finished class:  68\n",
      "Finished class:  69\n",
      "Finished class:  70\n",
      "Finished class:  71\n",
      "Finished class:  72\n",
      "Finished class:  73\n",
      "Finished class:  74\n",
      "Finished class:  75\n",
      "Finished class:  76\n",
      "Finished class:  77\n",
      "Finished class:  78\n",
      "Finished class:  79\n",
      "Finished class:  80\n",
      "Finished class:  81\n",
      "Finished class:  82\n",
      "Finished class:  83\n",
      "Finished class:  84\n",
      "Finished class:  85\n",
      "Finished class:  86\n",
      "Finished class:  87\n",
      "Finished class:  88\n",
      "Finished class:  89\n",
      "Finished class:  90\n",
      "Finished class:  91\n",
      "Finished class:  92\n",
      "Finished class:  93\n",
      "Finished class:  94\n",
      "Finished class:  95\n",
      "Finished class:  96\n",
      "Finished class:  97\n",
      "Finished class:  98\n",
      "Finished class:  99\n",
      "Finished class:  100\n",
      "2000000\n",
      "64_250_100_20000_Excl15_rand1\n",
      "Amino Acid exclusion:  [['S', 'V', 'A', 'N', 'T', 'D', 'R', 'Y', 'C', 'G', 'L', 'P', 'H', 'Amber/Stop', 'F'], ['H', 'T', 'E', 'Y', 'V', 'K', 'S', 'L', 'G', 'D', 'Q', 'I', 'C', 'R', 'P'], ['S', 'F', 'K', 'T', 'G', 'L', 'A', 'R', 'I', 'D', 'V', 'N', 'C', 'P', 'Amber/Stop'], ['K', 'A', 'L', 'Amber/Stop', 'P', 'D', 'T', 'Y', 'V', 'S', 'G', 'R', 'C', 'I', 'F'], ['V', 'R', 'K', 'G', 'L', 'S', 'A', 'P', 'D', 'I', 'E', 'N', 'C', 'Q', 'F'], ['S', 'P', 'A', 'E', 'I', 'T', 'Q', 'F', 'V', 'H', 'R', 'Y', 'G', 'N', 'Amber/Stop'], ['S', 'L', 'P', 'D', 'G', 'H', 'I', 'Q', 'F', 'R', 'W', 'V', 'N', 'Amber/Stop', 'A'], ['S', 'V', 'A', 'W', 'L', 'K', 'R', 'F', 'I', 'T', 'P', 'D', 'N', 'Y', 'H'], ['V', 'L', 'D', 'Q', 'S', 'R', 'G', 'E', 'Y', 'T', 'I', 'A', 'P', 'H', 'N'], ['Y', 'C', 'P', 'Amber/Stop', 'G', 'V', 'Q', 'T', 'D', 'L', 'A', 'R', 'S', 'H', 'I'], ['T', 'G', 'S', 'D', 'F', 'C', 'I', 'P', 'R', 'K', 'Amber/Stop', 'E', 'L', 'A', 'Q'], ['R', 'Q', 'T', 'I', 'V', 'N', 'P', 'C', 'Y', 'K', 'S', 'G', 'L', 'E', 'D'], ['H', 'G', 'K', 'V', 'P', 'T', 'A', 'S', 'C', 'D', 'E', 'R', 'I', 'L', 'F'], ['S', 'A', 'I', 'P', 'L', 'Y', 'W', 'H', 'N', 'E', 'F', 'Q', 'V', 'Amber/Stop', 'G'], ['P', 'D', 'L', 'Q', 'R', 'N', 'E', 'F', 'H', 'V', 'I', 'K', 'A', 'G', 'T'], ['S', 'V', 'A', 'E', 'R', 'I', 'P', 'T', 'Y', 'Amber/Stop', 'L', 'C', 'D', 'N', 'G'], ['V', 'R', 'F', 'T', 'H', 'D', 'E', 'P', 'I', 'L', 'Amber/Stop', 'K', 'N', 'C', 'S'], ['L', 'A', 'G', 'S', 'Q', 'I', 'R', 'C', 'T', 'V', 'P', 'N', 'E', 'F', 'H'], ['S', 'V', 'N', 'E', 'A', 'G', 'K', 'R', 'F', 'L', 'Amber/Stop', 'P', 'D', 'T', 'Y'], ['A', 'Q', 'S', 'G', 'F', 'R', 'L', 'Amber/Stop', 'P', 'K', 'V', 'N', 'D', 'E', 'H'], ['L', 'N', 'K', 'Q', 'H', 'C', 'Y', 'G', 'W', 'E', 'P', 'F', 'S', 'V', 'D'], ['S', 'Amber/Stop', 'P', 'G', 'Q', 'L', 'D', 'C', 'F', 'K', 'H', 'V', 'N', 'T', 'Y'], ['G', 'N', 'V', 'P', 'H', 'Q', 'Y', 'I', 'T', 'R', 'A', 'D', 'S', 'E', 'W'], ['E', 'A', 'P', 'S', 'Amber/Stop', 'R', 'L', 'Q', 'V', 'D', 'N', 'K', 'I', 'T', 'G'], ['S', 'G', 'A', 'K', 'Y', 'Amber/Stop', 'P', 'Q', 'H', 'L', 'R', 'I', 'T', 'V', 'D'], ['Y', 'E', 'R', 'L', 'Amber/Stop', 'N', 'P', 'I', 'G', 'A', 'Q', 'T', 'H', 'V', 'S'], ['R', 'L', 'F', 'A', 'W', 'N', 'S', 'G', 'I', 'V', 'E', 'D', 'Y', 'T', 'Q'], ['I', 'T', 'A', 'P', 'R', 'G', 'S', 'E', 'Y', 'D', 'C', 'L', 'V', 'W', 'N'], ['L', 'Y', 'T', 'S', 'H', 'F', 'D', 'V', 'R', 'P', 'G', 'Q', 'A', 'Amber/Stop', 'C'], ['R', 'D', 'G', 'V', 'P', 'Q', 'I', 'A', 'F', 'S', 'Amber/Stop', 'L', 'H', 'T', 'C'], ['G', 'R', 'S', 'D', 'C', 'W', 'L', 'Q', 'E', 'F', 'A', 'V', 'T', 'K', 'P'], ['F', 'S', 'T', 'P', 'G', 'L', 'Q', 'Amber/Stop', 'R', 'V', 'A', 'H', 'E', 'Y', 'K'], ['T', 'G', 'V', 'Q', 'K', 'P', 'A', 'N', 'I', 'R', 'S', 'L', 'W', 'H', 'Y'], ['L', 'R', 'K', 'A', 'I', 'T', 'C', 'G', 'Q', 'P', 'W', 'V', 'F', 'D', 'H'], ['E', 'A', 'I', 'R', 'K', 'Y', 'P', 'L', 'N', 'V', 'Q', 'H', 'F', 'S', 'G'], ['N', 'S', 'F', 'L', 'W', 'Y', 'V', 'Q', 'P', 'I', 'G', 'K', 'R', 'E', 'A'], ['T', 'A', 'R', 'N', 'V', 'H', 'K', 'I', 'G', 'W', 'C', 'Q', 'L', 'D', 'S'], ['C', 'V', 'E', 'Q', 'D', 'F', 'Amber/Stop', 'L', 'A', 'R', 'I', 'W', 'G', 'S', 'P'], ['S', 'P', 'K', 'A', 'V', 'G', 'Q', 'H', 'C', 'L', 'F', 'R', 'Y', 'T', 'D'], ['N', 'S', 'H', 'P', 'R', 'Y', 'D', 'Q', 'E', 'V', 'G', 'Amber/Stop', 'I', 'F', 'A'], ['G', 'E', 'A', 'L', 'K', 'R', 'I', 'Q', 'W', 'T', 'Y', 'H', 'C', 'S', 'Amber/Stop'], ['T', 'P', 'S', 'R', 'G', 'D', 'L', 'N', 'W', 'H', 'Y', 'C', 'V', 'Amber/Stop', 'I'], ['W', 'N', 'E', 'F', 'Q', 'C', 'G', 'H', 'A', 'L', 'I', 'S', 'V', 'P', 'R'], ['K', 'R', 'Amber/Stop', 'F', 'Q', 'A', 'P', 'I', 'L', 'G', 'N', 'V', 'T', 'Y', 'S'], ['F', 'R', 'L', 'H', 'G', 'I', 'P', 'N', 'A', 'W', 'E', 'V', 'T', 'S', 'D'], ['S', 'R', 'P', 'L', 'V', 'D', 'I', 'A', 'N', 'Q', 'F', 'T', 'Y', 'H', 'G'], ['Amber/Stop', 'L', 'P', 'T', 'K', 'S', 'R', 'G', 'Y', 'A', 'E', 'F', 'D', 'I', 'C'], ['R', 'Y', 'H', 'A', 'I', 'P', 'D', 'V', 'G', 'C', 'T', 'S', 'L', 'W', 'Amber/Stop'], ['L', 'D', 'P', 'K', 'V', 'G', 'T', 'I', 'A', 'F', 'N', 'R', 'E', 'C', 'Y'], ['T', 'P', 'L', 'E', 'R', 'Amber/Stop', 'F', 'G', 'S', 'A', 'K', 'V', 'D', 'I', 'W'], ['T', 'N', 'S', 'K', 'Q', 'V', 'G', 'R', 'I', 'F', 'P', 'L', 'Y', 'D', 'A'], ['I', 'A', 'D', 'Amber/Stop', 'P', 'T', 'L', 'R', 'G', 'C', 'S', 'N', 'V', 'Q', 'H'], ['V', 'A', 'D', 'R', 'I', 'S', 'L', 'F', 'T', 'P', 'G', 'E', 'C', 'N', 'H'], ['T', 'Q', 'S', 'R', 'H', 'C', 'V', 'I', 'F', 'D', 'A', 'L', 'P', 'E', 'G'], ['Y', 'T', 'V', 'G', 'I', 'L', 'P', 'H', 'C', 'K', 'R', 'F', 'Q', 'S', 'N'], ['S', 'P', 'L', 'V', 'R', 'T', 'Y', 'C', 'D', 'H', 'E', 'I', 'F', 'K', 'N'], ['Y', 'V', 'P', 'R', 'L', 'I', 'S', 'A', 'Q', 'T', 'C', 'G', 'E', 'N', 'D'], ['Y', 'L', 'R', 'N', 'P', 'Q', 'G', 'V', 'T', 'F', 'I', 'H', 'D', 'A', 'E'], ['Q', 'L', 'A', 'N', 'S', 'R', 'V', 'H', 'E', 'I', 'T', 'W', 'D', 'G', 'Y'], ['Q', 'A', 'V', 'S', 'L', 'R', 'T', 'G', 'E', 'C', 'Y', 'P', 'D', 'K', 'Amber/Stop'], ['W', 'R', 'A', 'V', 'D', 'K', 'L', 'S', 'P', 'H', 'I', 'T', 'C', 'Y', 'Q'], ['S', 'R', 'L', 'T', 'A', 'P', 'K', 'V', 'C', 'E', 'G', 'Y', 'Q', 'I', 'F'], ['R', 'Y', 'T', 'V', 'C', 'F', 'A', 'L', 'H', 'I', 'K', 'P', 'S', 'G', 'E'], ['L', 'F', 'Y', 'T', 'G', 'Amber/Stop', 'K', 'S', 'V', 'C', 'W', 'E', 'P', 'A', 'H'], ['V', 'S', 'R', 'G', 'F', 'C', 'A', 'H', 'Q', 'W', 'P', 'L', 'E', 'I', 'D'], ['S', 'N', 'V', 'D', 'P', 'R', 'T', 'H', 'C', 'Q', 'G', 'L', 'A', 'K', 'Y'], ['E', 'D', 'R', 'N', 'V', 'K', 'S', 'I', 'A', 'L', 'Q', 'T', 'P', 'Y', 'Amber/Stop'], ['R', 'T', 'A', 'S', 'Y', 'P', 'I', 'D', 'L', 'Amber/Stop', 'K', 'Q', 'G', 'C', 'F'], ['C', 'P', 'R', 'Y', 'N', 'A', 'K', 'V', 'S', 'W', 'T', 'L', 'Amber/Stop', 'F', 'G'], ['L', 'A', 'Q', 'R', 'H', 'G', 'S', 'P', 'T', 'Amber/Stop', 'V', 'E', 'Y', 'N', 'C'], ['I', 'H', 'S', 'F', 'L', 'V', 'G', 'Q', 'T', 'R', 'D', 'N', 'C', 'Amber/Stop', 'P'], ['D', 'P', 'L', 'S', 'A', 'I', 'N', 'Amber/Stop', 'R', 'T', 'V', 'Y', 'C', 'F', 'W'], ['S', 'R', 'Q', 'D', 'C', 'E', 'K', 'L', 'P', 'H', 'G', 'V', 'F', 'A', 'I'], ['F', 'R', 'V', 'S', 'G', 'A', 'T', 'L', 'D', 'Q', 'H', 'I', 'Y', 'C', 'W'], ['S', 'E', 'C', 'R', 'P', 'I', 'F', 'G', 'Y', 'A', 'H', 'N', 'L', 'Q', 'T'], ['S', 'Q', 'L', 'D', 'C', 'K', 'G', 'R', 'N', 'P', 'V', 'A', 'Amber/Stop', 'T', 'W'], ['S', 'D', 'F', 'G', 'Q', 'P', 'V', 'L', 'W', 'E', 'N', 'H', 'I', 'R', 'T'], ['V', 'P', 'R', 'D', 'Q', 'L', 'E', 'C', 'S', 'T', 'G', 'K', 'A', 'F', 'Y'], ['P', 'R', 'A', 'C', 'L', 'Amber/Stop', 'Q', 'F', 'V', 'E', 'I', 'D', 'T', 'G', 'N'], ['K', 'R', 'T', 'S', 'H', 'N', 'E', 'Amber/Stop', 'G', 'L', 'V', 'F', 'I', 'C', 'W'], ['K', 'A', 'R', 'S', 'D', 'T', 'G', 'N', 'F', 'Q', 'P', 'L', 'H', 'I', 'C'], ['R', 'F', 'G', 'S', 'W', 'P', 'K', 'A', 'V', 'N', 'I', 'Amber/Stop', 'D', 'T', 'Q'], ['V', 'H', 'F', 'A', 'G', 'I', 'D', 'R', 'K', 'P', 'C', 'N', 'T', 'L', 'S'], ['H', 'A', 'Y', 'R', 'I', 'L', 'S', 'E', 'G', 'Q', 'T', 'W', 'N', 'V', 'C'], ['A', 'P', 'C', 'R', 'E', 'Q', 'N', 'H', 'F', 'I', 'L', 'D', 'S', 'T', 'G'], ['V', 'G', 'S', 'R', 'T', 'L', 'C', 'A', 'D', 'P', 'H', 'Amber/Stop', 'W', 'Y', 'F'], ['L', 'A', 'V', 'G', 'R', 'Amber/Stop', 'D', 'Q', 'Y', 'I', 'T', 'C', 'N', 'S', 'P'], ['L', 'A', 'S', 'T', 'G', 'V', 'H', 'R', 'D', 'I', 'C', 'P', 'K', 'Y', 'Q'], ['K', 'Amber/Stop', 'P', 'L', 'V', 'T', 'A', 'N', 'Q', 'D', 'S', 'F', 'Y', 'H', 'C'], ['C', 'G', 'R', 'S', 'V', 'W', 'Y', 'T', 'F', 'P', 'Amber/Stop', 'N', 'H', 'L', 'I'], ['S', 'F', 'L', 'T', 'Q', 'C', 'V', 'H', 'D', 'P', 'R', 'G', 'N', 'I', 'A'], ['T', 'D', 'A', 'Amber/Stop', 'C', 'S', 'V', 'G', 'P', 'N', 'Q', 'L', 'H', 'K', 'R'], ['Q', 'P', 'E', 'V', 'R', 'C', 'G', 'I', 'S', 'L', 'F', 'T', 'Amber/Stop', 'A', 'W'], ['R', 'V', 'N', 'P', 'A', 'Y', 'L', 'I', 'G', 'T', 'Amber/Stop', 'W', 'Q', 'C', 'S'], ['L', 'F', 'S', 'A', 'K', 'T', 'R', 'G', 'E', 'P', 'C', 'I', 'D', 'Y', 'H'], ['S', 'T', 'R', 'A', 'G', 'Q', 'K', 'H', 'P', 'D', 'Y', 'L', 'I', 'E', 'V'], ['F', 'L', 'P', 'A', 'I', 'R', 'G', 'E', 'K', 'N', 'T', 'C', 'V', 'S', 'Y'], ['Q', 'Y', 'V', 'R', 'F', 'A', 'T', 'C', 'E', 'P', 'G', 'I', 'L', 'S', 'D'], ['Q', 'R', 'H', 'Y', 'V', 'L', 'E', 'G', 'T', 'K', 'S', 'Amber/Stop', 'A', 'D', 'F'], ['S', 'C', 'R', 'Y', 'F', 'G', 'I', 'V', 'P', 'A', 'E', 'L', 'Q', 'D', 'N']]\n"
     ]
    }
   ],
   "source": [
    "num_aminos = 64\n",
    "seq_length = 250\n",
    "num_classes = 100\n",
    "seq_per_class = 20000\n",
    "keys, val, Excl = create_set(num_aminos, seq_length, num_classes, seq_per_class)\n",
    "print(len(keys))\n",
    "#print(len(val))\n",
    "#list of tuples (x,y) where x is a string and y is index\n",
    "fout = str(num_aminos) + '_' + str(seq_length) + '_' + str(num_classes) + '_' + str(seq_per_class) + '_Excl15_rand1'\n",
    "print(fout)\n",
    "print(\"Amino Acid exclusion: \", Excl)\n",
    "with open('../data/data%s.csv' % fout, 'w') as f:\n",
    "#with open('data/64_30_30_10.csv', 'w') as f:\n",
    "    writer = csv.writer(f)\n",
    "    writer.writerow(['class', 'sequence'])\n",
    "    for i in range(0, len(keys)):\n",
    "        writer.writerow([keys[i], val[i]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
